File: src/main/java/org/mitallast/queue/crdt/registry/CrdtRegistry.java
package org.mitallast.queue.crdt.registry;

import javaslang.control.Option;
import org.mitallast.queue.crdt.Crdt;

public interface CrdtRegistry {

    int index();

    boolean createLWWRegister(long id);

    boolean remove(long id);

    Crdt crdt(long id);

    Option<Crdt> crdtOpt(long id);

    <T extends Crdt> T crdt(long id, Class<T> type);

    <T extends Crdt> Option<T> crdtOpt(long id, Class<T> type);
}

File: src/main/java/org/mitallast/queue/crdt/registry/CrdtRegistry.kt
package org.mitallast.queue.crdt.registry

import io.vavr.control.Option
import org.mitallast.queue.crdt.Crdt

interface CrdtRegistry {

    fun index(): Int

    fun createLWWRegister(id: Long): Boolean

    fun createGCounter(id: Long): Boolean

    fun createGSet(id: Long): Boolean

    fun createOrderedGSet(id: Long): Boolean

    fun remove(id: Long): Boolean

    fun crdt(id: Long): Crdt

    fun crdtOpt(id: Long): Option<Crdt>

    fun <T : Crdt> crdt(id: Long, type: Class<T>): T

    fun <T : Crdt> crdtOpt(id: Long, type: Class<T>): Option<T>
}
